attestation-service/attestation-service/src/policy_engine/opa/mod.rs

```diff
@@ -63,51 +63,74 @@ impl PolicyEngine for OPA {
+                use sha2::Digest;
+                let mut hasher = sha2::Sha384::new();
+                hasher.update(&policy);
+                let hex = hasher.finalize().to_vec();
+                hex::encode(hex)
+            };
+
+            let policy_go = GoString {
+                p: policy.as_ptr() as *const c_char,
+                n: policy.len() as isize,
+            };
+
+            let reference = serde_json::json!({ "reference": reference_data_map }).to_string();
+
+            let reference_go = GoString {
+                p: reference.as_ptr() as *const c_char,
+                n: reference.len() as isize,
+            };
+
+            let input_go = GoString {
+                p: input.as_ptr() as *const c_char,
+                n: input.len() as isize,
+            };
+
+            // Call the function exported by cgo and process the returned decision
+            let decision_buf: *mut c_char =
+                unsafe { evaluateGo(policy_go, reference_go, input_go) };
+            let decision_str: &CStr = unsafe { CStr::from_ptr(decision_buf) };
+            let policy_res = decision_str.to_str()?.to_string();
+            debug!("Evaluated: {}", policy_res);
+            if policy_res.starts_with("Error::") {
+                bail!("OPA verification failed: {policy_res}");
+            }
 
-        let input_go = GoString {
-            p: input.as_ptr() as *const c_char,
-            n: input.len() as isize,
-        };
+            // If a clear approval opinion is given in the evaluation report,
+            // the rejection information will be reflected in the evaluation failure return value.
+            let res_kv: Value = serde_json::from_str(&policy_res)?;
 
-        // Call the function exported by cgo and process the returned decision
-        let decision_buf: *mut c_char = unsafe { evaluateGo(policy_go, reference_go, input_go) };
-        let decision_str: &CStr = unsafe { CStr::from_ptr(decision_buf) };
-        let res = decision_str.to_str()?.to_string();
-        debug!("Evaluated: {}", res);
-        if res.starts_with("Error::") {
-            return Err(anyhow!(res));
-        }
+            // only if there is a field named `allow` in the evaluation report and
+            // it is false, the evaluation fails. Otherwise the evaluation will be
+            // treated as succees.
+            let allow = res_kv
+                .get("allow")
+                .map(|a| a.as_bool())
+                .map(|a| a.unwrap_or(true))
+                .unwrap_or(true);
 
-        // If a clear approval opinion is given in the evaluation report,
-        // the rejection information will be reflected in the evaluation failure return value.
-        let res_kv: Value = serde_json::from_str(&res)?;
-        if let Some(allow) = res_kv["allow"].as_bool() {
             if !allow {
-                bail!("Untrusted TEE evidence")
+                bail!("TEE evidence does not pass policy {policy_id}");
```

Comment:
does the policy evaluation return a more verbose error when evaluation fails? something like "field X didn't match condition Y".

if so, then let's return that.